client/server/middleware/assets.js

```diff
@@ -29,8 +28,7 @@ const groupAssetsByType = ( assets ) => defaults( groupBy( assets, getAssetType
 export default () => {
 	return ( req, res, next ) => {
 		req.getAssets = () => {
-			const target = req.getTarget();
-			return JSON.parse( fs.readFileSync( getAssetsPath( target ), 'utf8' ) );
+			return JSON.parse( fs.readFileSync( getAssetsPath(), 'utf8' ) );
```

Comment:
This `fs.readFileSync` is something that should be fixed:
1. The file is re-read on every request, three times: in `getFilesForEntrypoint()`, in `getFilesForChunk()` and in `request.getAssets().manifests`. It should be read only once at server boot and then cached in memory.
2. The read is sync, and that means blocking the event loop, limiting the server parallelism. The read should be async, postponing the `next()` call until the file is read.

Something like:
```js
let assetsFile;
function readAssets() {
  if ( ! assetsFile ) {
    assetsFile = fs.promises.readFile( path ).then( text => JSON.parse( text ) );
  }
  return assetsFile;
}

const midware = async ( req, res, next ) => {
  const assets = await readAssets();
  Object.assign( req, {
    getAssets: () => assets,
    getFilesForEntrypoint: ( name ) => {
      const entrypointAssets = assets.entrypoints[ name ].assets;
      /* ... */
    }
  } );
  next();
};
```

Reply:
I agree this should be done outside each request to save time.

What do you think about keeping it sync and having it at the middleware constructor? Something like

```
//assets.js
export default () => {
	const assets = fs.readFileSync(/*...*/);
	return ( req, res, next ) => {
		Object.assign( req, {
			getAssets: () => assets,
			/* ... */
		} );
	};
};

//index.js
import middlewareAssets from '../middleware/assets.js';
/*...*/
app.use( middlewareAssets() );
```

My reasoning is that if reading the assets file throws (e.g. maybe it doesn't exist) that's a critical error and the server should just fail to start in that scenario. I think that's better than the server starting, our infrastructure thinking the deployment was ok and then failing on the very first request because there are assets file.

Reply:
> What do you think about keeping it sync and having it at the middleware constructor?

You can try it but I suspect that it might fail in development mode. In dev mode, we first start the server and then the server internally starts webpack, which writes the `assets-evergreen.json` file. So, at the time when the middleware is initialized, the assets file doesn't exist yet.

Or is the assets file generated and used only in production mode? I'm not sure. But it's something that needs to be checked when going this way.

I agree that reading the assets file once at server boot and terminating when it's not found is preferable.

Reply:
> You can try it but I suspect that it might fail in development mode

You're 100% right :( The file doesn't exist when the server start in dev mode.

Reply:
Done.